test/test_autograd.py

```diff
@@ -3006,6 +3006,29 @@ def parent_on_cpu(inp):
         # gpu thread ReadyQueue
         out.sum().backward()
 
+    def test_set_data_preserve_version_counter(self):
+        x = torch.Tensor(1, 2)
+
+        # In-place op bumps version
+        x_saved_version = x._version
+        x.add_(1).add_(1)
+        self.assertTrue(x._version > x_saved_version)
+
+        # Differentiable view shares version counter
+        xz = x[:]
+        self.assertTrue(x._version == xz._version)
+
+        # `x.data = y` preserves version counter of `x`
+        x_saved_version = x._version
+        x.data = torch.Tensor(2, 3)
+        self.assertTrue(x._version == x_saved_version)
+        self.assertTrue(x._version == xz._version)
+
+        # In-place op on `x` also updates version of `xz`,
+        # because they share the version counter
+        x.add_(1)
+        self.assertTrue(x._version == xz._version)
```

Comment:
Some improvements to this test:

1) change the variable through xz.
2) check that changes post-set data actually increase the version counter.